File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.lua
local function DecimalParaHexadecimal(decimal)
    local resto
    local hexadecimal = ""
    repeat
        resto = decimal % 16
        if resto >= 0 and resto <= 9 then
            hexadecimal = tostring(resto) .. hexadecimal
        elseif resto >= 10 and resto <= 15 then
            hexadecimal = string.char(resto - 10 + string.byte("A")) .. hexadecimal
        end
        decimal = math.floor(decimal / 16)
    until decimal == 0

    return hexadecimal
end

local function HexadecimalParaDecimal(hexadecimal)
    local potencia = 0
    local decimal = 0
    while #hexadecimal > 0 do
        local digito = string.upper(string.sub(hexadecimal, #hexadecimal, #hexadecimal))
        local valor
        if digito >= "0" and digito <= "9" then
            valor = tonumber(digito)
        elseif digito >= "A" and digito <= "F" then
            valor = string.byte(digito) - string.byte("A") + 10
        else
            return nil
        end
        decimal = decimal + valor * (16 ^ potencia)
        potencia = potencia + 1
        hexadecimal = string.sub(hexadecimal, 1, #hexadecimal - 1)
    end
    return decimal
end

local function IsValidHexadecimal(hexadecimal)
    return string.match(hexadecimal, "^[0-9A-Fa-f]+$") ~= nil
end

local LimiteMaximoDecimal = 2147483647
local LimiteMaximoHexadecimal = "7FFFFFFF"

repeat
    print("===============================")
    print(" CONVERSOR DECIMAL-HEXADECIMAL ")
    print("===============================")
    print()
    print("1 - Converter decimal para hexadecimal")
    print("2 - Converter hexadecimal para decimal")
    print("0 - Sair")
    print()
    io.write("Digite a opcao desejada: ")
    local opcao = tonumber(io.read())
    if opcao == 1 then
        print()
        io.write("Digite um numero decimal (ate " .. LimiteMaximoDecimal .. "): ")
        local decimal = tonumber(io.read())
        if decimal and decimal >= 0 and decimal <= LimiteMaximoDecimal then
            local hexadecimal = DecimalParaHexadecimal(decimal)
            print("Valor em hexadecimal: " .. hexadecimal)
        else
            print("Valor decimal fora do limite permitido.")
        end
    elseif opcao == 2 then
        print()
        io.write("Digite um numero hexadecimal (ate " .. LimiteMaximoHexadecimal .. "): ")
        local hexadecimal = io.read()
        if IsValidHexadecimal(hexadecimal) then
            local decimal = HexadecimalParaDecimal(hexadecimal)
            if decimal and decimal <= LimiteMaximoDecimal then
                print("Valor em decimal: " .. math.floor(decimal))
            else
                print("Valor hexadecimal fora do limite permitido.")
            end
        else
            print("Valor hexadecimal invalido.")
        end
    end
    print()
until opcao == 0

File: Conversoes de Bases/Decimal e Hexadecimal/DecimalHexadecimal.pas
program DecimalHexadecimal;

uses
  SysUtils, Math;

const
  LimiteMaximoDecimal: Cardinal = 2147483647;
  LimiteMaximoHexadecimal: string = '7FFFFFFF';

function DecimalParaHexadecimal(decimal: Cardinal): string;
var
  resto: Cardinal;
  hexadecimal: string;
begin
  hexadecimal := '';
  repeat
    resto := decimal mod 16;
    case resto of
      0..9:
        hexadecimal := Chr(resto + Ord('0')) + hexadecimal;
      10..15:
        hexadecimal := Chr(resto - 10 + Ord('A')) + hexadecimal;
    end;
    decimal := decimal div 16;
  until decimal = 0;
  DecimalParaHexadecimal := hexadecimal;
end;

function HexadecimalParaDecimal(hexadecimal: string): Cardinal;
var
  potencia: Cardinal;
  digito: Char;
  decimal: Cardinal;
begin
  decimal := 0;
  potencia := 0;
  while Length(hexadecimal) > 0 do
  begin
    digito := UpCase(hexadecimal[Length(hexadecimal)]);
    case digito of
      '0'..'9':
        decimal := decimal + (Ord(digito) - Ord('0')) * Trunc(Power(16, potencia));
      'A'..'F':
        decimal := decimal + (Ord(digito) - Ord('A') + 10) * Trunc(Power(16, potencia));
    end;
    Delete(hexadecimal, Length(hexadecimal), 1);
    potencia := potencia + 1;
  end;

  HexadecimalParaDecimal := decimal;
end;

function IsValidHexadecimal(hexadecimal: string): Boolean;
var
  i: Integer;
begin
  for i := 1 to Length(hexadecimal) do
  begin
    if not (UpCase(hexadecimal[i]) in ['0'..'9', 'A'..'F']) then
    begin
      IsValidHexadecimal := False;
      Exit;
    end;
  end;
  IsValidHexadecimal := True;
end;

var
  opcao: Integer;
  decimal: Cardinal;
  hexadecimal: string;

begin
  repeat
    Writeln('===============================');
    Writeln(' CONVERSOR DECIMAL-HEXADECIMAL ');
    Writeln('===============================');
    Writeln;
    Writeln('1 - Converter decimal para hexadecimal');
    Writeln('2 - Converter hexadecimal para decimal');
    Writeln('0 - Sair');
    Writeln;
    Write('Digite a opcao desejada: ');
    Readln(opcao);
    case opcao of
      1:
      begin
        WriteLn;
        Write('Digite um numero decimal (ate ', LimiteMaximoDecimal, '): ');
        Readln(decimal);
        if decimal > LimiteMaximoDecimal then
          Writeln('Valor decimal fora do limite permitido.')
        else
          Writeln('Valor em hexadecimal: ', DecimalParaHexadecimal(decimal));
      end;
      2:
      begin
        WriteLn;
        Write('Digite um numero hexadecimal (ate ', LimiteMaximoHexadecimal, '): ');
        Readln(hexadecimal);
        if not IsValidHexadecimal(hexadecimal) then
          Writeln('Valor hexadecimal invalido.')
        else if HexadecimalParaDecimal(hexadecimal) > LimiteMaximoDecimal then
          Writeln('Valor hexadecimal fora do limite permitido.')
        else
          Writeln('Valor em decimal: ', HexadecimalParaDecimal(hexadecimal));
      end;
    end;
    Writeln;
  until opcao = 0;
end.
